### PR TITLE
Add status filters for Sessions

### PR DESCRIPTION
**Pull request by andresag4**
## Description

Allow User to filter by `Past` `Live` and `Starting Soon` status.

## How has this been tested?

Please mark the tests that you ran to verify your changes. If difficult to test, consider providing instructions so reviewers can test.

- [x] Manual testing
- [ ] System tests
- [x] Unit tests
- [ ] None

## Checklist

- [x] CI pipeline is passing
- [x] My code follows the conventions of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added seed data to the database (if applicable)

## Release tasks

Add any tasks that need to be done before/after the release of this feature.

## Screenshots/Loom

image
image

### CODE REVIEW

File: app/queries/session_query.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+class SessionQuery
```
Comment (EdgarB) [inline app/queries/session_query.rb:1]: Is it necessary to have this Session Query? Seems to be a bit over-engineering

Also, I think here is a clearer way to define the query object. Removing the scopes from the model and passing them to the query object
https://thoughtbot.com/blog/a-case-for-query-objects-in-rails

File: app/views/sessions/index.html.erb [hunk 2]
```diff
@@ -6,22 +6,81 @@
...
+                      <%= form.label name do %>
+                        <%= t("views.status_filters.#{name}") %>
+                      <% end %>
+                    </div>
+                  </div>
+                <% end %>
+                <div class="h-6 px-2.5 py-4 rounded justify-center items-center gap-2.5 inline-flex">
+                  <div class="text-lg text-red italic underline font-black">
+                    <%= form.submit t("views.status_filters.submit"), name: "", class: "text-lg text-red italic underline font-black" %>
+                  </div>
```
Comment (EdgarB) [inline app/views/sessions/index.html.erb:47]: We need to update the dropdown shadow, items font weight and remove the button in order to match the designs;
Screenshot 2024-08-13 at 6 17 12 p m
Screenshot 2024-08-13 at 6 17 06 p m

Comment (EdgarB) [reply app/views/sessions/index.html.erb:47]: We also need to position the dropdown a bit more to the right and bottom to match the designs.

In order to submit the filters after a checkbox change you could use the "form" stimulus controller.

Also, not a blocker (nice to have) the filters button is not filling with the red background when the dropdown is opened

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -1,9 +1,13 @@
...
     if request.params[:on_date] == session_date.to_s || (session_date.blank? && request.params[:on_date].blank?)
       "bg-red text-white"
     else
       "bg-white text-red"
     end
   end
+
+  def active_filters?
+    SessionQuery::STATUS_SCOPES.any? { |status| request.params[status] == "1" }
+  end
```
Comment (EdgarB) [inline app/helpers/application_helper.rb:12]: Could we move this to a `SessionsHelper` module and keep the method names scoped to "sessions"? (given that they are global)

File: app/helpers/session_helper.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+module SessionHelper
```
Comment (Sergio-e) [inline app/helpers/session_helper.rb:1]: Helpers are available on every page, which means having generic method names will become a problem as the application starts growing. I suggest picking a name that relates to the domain of the helper module

File: app/models/session.rb [hunk 5]
```diff
@@ -32,9 +32,25 @@ class Session < ApplicationRecord
 
   validates_datetime :ends_at, after: :starts_at
 
-  scope :starts_at, ->(date) { where(starts_at: date.all_day) }
+  scope :starts_at, ->(date) { where("date(starts_at) = ?", date) }
+  scope :past, -> { where(ends_at: ...Time.current) }
+  scope :live, -> { where("? BETWEEN starts_at AND ends_at", Time.current) }
+  scope :starting_soon, -> { where("starts_at BETWEEN ? and ?", Time.current, 1.hour.from_now) }
+  scope :from_user, ->(user) { joins(:attendees).where(attendees: {id: user.id}) }
```
Comment (Sergio-e) [inline app/models/session.rb:39]:
```suggestion
```

File: app/views/sessions/_filters.html.erb [hunk 6]
```diff
@@ -3,28 +3,87 @@
...
+      <div
+        data-dropdown-target="menu"
+        class="hidden transition transform origin-top-right absolute right-0"
+        data-transition-enter-from="opacity-0 scale-95"
+        data-transition-enter-to="opacity-100 scale-100"
+        data-transition-leave-from="opacity-100 scale-100"
+        data-transition-leave-to="opacity-0 scale-95">
+        <div class="pt-2 pb-2 bg-white rounded-lg shadow-[0_10px_50px_0_rgba(0,0,0,0.40)] flex-col inline-flex absolute top-4 right-[-8px] right-0 w-40 max-w-40">
+          <%= form_with url: sessions_path, class: "grid", method: :get do |form| %>
+            <%= form.hidden_field :starts_at, value: params[:starts_at] %>
```
Comment (Sergio-e) [inline app/views/sessions/_filters.html.erb:42]: let's use a hidden_field to maintain the active filters when querying over a different date. Currently, the active filters will be lost if a new date is selected
image

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -1,7 +1,9 @@
 class SessionsController < ApplicationController
   def index
-    @sessions = sessions.includes(:attendees, :speakers, :location, :tags)
-    @sessions = @sessions.starts_at(params[:starts_at].to_date) if params[:starts_at].present?
+    @sessions = SessionQuery.new(
```
Comment (Sergio-e) [inline app/controllers/sessions_controller.rb:3]: I think we need to use the same query object on the `SchedulesController`

File: spec/queries/session_query_spec.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+RSpec.describe SessionQuery, type: :query do
```
Comment (Sergio-e) [inline spec/queries/session_query_spec.rb:5]: 👍

File: app/javascript/controllers/index.js [hunk 9]
```diff
@@ -4,8 +4,11 @@ import { application } from 'controllers/application'
...
 import { eagerLoadControllersFrom } from '@hotwired/stimulus-loading'
-eagerLoadControllersFrom('controllers', application)
 
 // Lazy load controllers as they appear in the DOM (remember not to preload controllers in import map!)
 // import { lazyLoadControllersFrom } from "@hotwired/stimulus-loading"
 // lazyLoadControllersFrom("controllers", application)
+
+import Dropdown from '@stimulus-components/dropdown'
+eagerLoadControllersFrom('controllers', application)
+application.register('dropdown', Dropdown)
```
Comment (LuigiR0jas) [inline app/javascript/controllers/index.js:14]:
```suggestion
import Dropdown from '@stimulus-components/dropdown'

// Lazy load controllers as they appear in the DOM (remember not to preload controllers in import map!)
// import { lazyLoadControllersFrom } from "@hotwired/stimulus-loading"
// lazyLoadControllersFrom("controllers", application)
eagerLoadControllersFrom('controllers', application)
application.register('dropdown', Dropdown)
```

File: app/models/application_record.rb [hunk 10]
```diff
@@ -1,3 +1,18 @@
 class ApplicationRecord < ActiveRecord::Base
   primary_abstract_class
+
+  class << self
```
Comment (LuigiR0jas) [inline app/models/application_record.rb:4]: Why is this needed?

File: app/views/sessions/_card.html.erb [hunk 11]
```diff
@@ -3,6 +3,17 @@
 
 <div class="flex flex-col w-full p-5 mb-4 bg-white rounded-[20px]">
   <div class="flex flex-col w-full">
+    <% if session.live? %>
```
Comment (LuigiR0jas) [inline app/views/sessions/_card.html.erb:6]: I think we're missing the past and upcoming session status label

File: app/models/application_record.rb (hunk 10 above)
Comment (andresag4) [reply app/models/application_record.rb:4]: prefer to delcare this insted of `def self.method_name`

File: app/controllers/sessions_controller.rb [hunk 12]
```diff
@@ -1,11 +1,7 @@
 class SessionsController < ApplicationController
   def index
-    @user_session_ids = current_user.sessions.pluck(:id)
-    @sessions = sessions
-      .joins(:speakers, :location)
+    @sessions = SessionQuery.new(relation: Session.joins(:speakers, :location), params: filter_params).call
```
Comment (LuigiR0jas) [inline app/controllers/sessions_controller.rb:3]: We're missing @user_session_ids

File: app/views/sessions/_filters.html.erb [hunk 13]
```diff
@@ -3,28 +3,87 @@
...
+
+<% if active_session_filters? %>
+  <div class="flex w-full mb-2 gap-2">
+    <div class="flex flex-wrap ">
+      <div class="h-8 py-1 text-base font-bold leading-tight">Active filters:</div>
+    </div>
+    <div class="flex flex-wrap gap-2">
+      <% SessionQuery::STATUS_SCOPES.each do |name| %>
+        <% next unless params[name] == "1" %>
+        <div class="h-8 px-4 py-2 bg-green-500 bg-opacity-[.08] rounded-lg gap-2 inline-flex">
```
Comment (LuigiR0jas) [inline app/views/sessions/_filters.html.erb:74]: The style of each filter should be different
Screenshot 2024-08-15 at 3 23 58 p m

File: app/views/sessions/_filters.html.erb [hunk 14]
```diff
@@ -3,28 +3,87 @@
...
+          ) %>
     <% end %>
   </div>
 
-  <div class="flex items-center justify-center ml-3 p-2 bg-white rounded-full min-w-[40px] min-h-[40px]">
-    <%= inline_svg_tag("icons/filters.svg", size: "18") %>
+  <div class="flex items-center justify-center ml-3 p-2 rounded-full min-w-[40px] min-h-[40px] <%= active_session_filters? ? "bg-red text-white" : "bg-white text-red" %>">
+    <div class="relative" data-controller="dropdown">
+      <button type="button" data-action="dropdown#toggle click@window->dropdown#hide" class="flex items-center justify-center p-2 rounded-full">
+        <%= inline_svg_tag("icons/filters.svg", size: "18") %>
```
Comment (LuigiR0jas) [inline app/views/sessions/_filters.html.erb:30]: When filters are applied, the filter icon sh
Screenshot 2024-08-15 at 3 24 40 p m
ould turn white

File: app/views/sessions/_filters.html.erb [hunk 15]
```diff
@@ -3,28 +3,87 @@
...
+              </div>
+            </div>
+          <% end %>
+        </div>
+      </div>
+    </div>
   </div>
 </div>
+
+<% if active_session_filters? %>
```
Comment (LuigiR0jas) [inline app/views/sessions/_filters.html.erb:66]: The empty state when no sessions match the current filters should be this one
Screenshot 2024-08-15 at 3 26 09 p m

File: app/views/sessions/_card.html.erb [hunk 16]
```diff
@@ -2,6 +2,17 @@
...
+      <div class="h-8 px-4 py-2 bg-green-500 rounded-lg bg-opacity-[.08] justify-center items-center inline-flex mb-4">
+        <div class="grow shrink basis-0 h-4 justify-center items-center gap-2 flex">
+          <span class="relative flex h-4 w-4">
+            <span class="animate-slow-ping absolute inline-flex h-full w-full rounded-full bg-green-500 opacity-75"></span>
+            <span class="relative inline-flex rounded-full h-4 w-4 bg-green-500"></span>
+          </span>
+          <div class="grow shrink">Session is live now.</div>
```
Comment (LuigiR0jas) [inline app/views/sessions/_card.html.erb:12]: Couple of improvements:
1. No need to use flex grow and shrink. It should be enough with w-full for both the child and the flex parent, and use text-align: left.
2. This text seems a little bit bigger than in the design. Also, use an appropriate tag for text of this kind, such as p or span

File: app/views/sessions/_card.html.erb [hunk 17]
```diff
@@ -2,6 +2,17 @@
 
 <div class="flex flex-col w-full p-5 mb-4 bg-white rounded-[20px]">
   <div class="flex flex-col w-full">
+    <% if session.live? %>
+      <div class="h-8 px-4 py-2 bg-green-500 rounded-lg bg-opacity-[.08] justify-center items-center inline-flex mb-4">
+        <div class="grow shrink basis-0 h-4 justify-center items-center gap-2 flex">
+          <span class="relative flex h-4 w-4">
+            <span class="animate-slow-ping absolute inline-flex h-full w-full rounded-full bg-green-500 opacity-75"></span>
+            <span class="relative inline-flex rounded-full h-4 w-4 bg-green-500"></span>
```
Comment (LuigiR0jas) [inline app/views/sessions/_card.html.erb:10]: No need to use relative or inline-flex, just use regular flex, width: 100% and justify-content or align-items as per necessary

File: app/views/sessions/_card.html.erb (hunk 11 above)
Comment (LuigiR0jas) [reply app/views/sessions/_card.html.erb:5]: Note: Their icon should not pulse

File: app/views/sessions/_card.html.erb (hunk 16 above)
Comment (andresag4) [reply app/views/sessions/_card.html.erb:12]: had to keep grow on parent
but child is now w-full

File: app/models/session.rb [hunk 18]
```diff
@@ -32,9 +32,24 @@ class Session < ApplicationRecord
...
 
   def self.ransackable_attributes(_auth_object = nil)
     %w[title]
   end
+
+  def live?
+    Time.current.between?(starts_at, ends_at)
+  end
+
+  def starting_soon?
```
Comment (Sergio-e) [inline app/models/session.rb:48]: Doesn't seem to be correct. I created a session that `starts_at` on Aug 13 and ends on Aug14 and it is being marked as "starting soon"
image

File: app/views/sessions/_filters.html.erb (hunk 6 above)
Comment (Sergio-e) [reply app/views/sessions/_filters.html.erb:42]: Reminder - this one needs to be addressed

File: app/views/sessions/index.html.erb [hunk 19]
```diff
@@ -17,7 +17,7 @@
             }
           ) %>
         <% end %>
-      <% elsif params[:starts_at].present? %>
+      <% elsif params[:starts_at].present? || active_session_filters? %>
```
Comment (Sergio-e) [inline app/views/sessions/index.html.erb:20]: I suggest making `filter_params` a [helper_method](https://api.rubyonrails.org/classes/AbstractController/Helpers/ClassMethods.html#method-i-helper_method) and using `filter_params.present?`. This same change should be applied to schedules#show
